support regex tag filter on master

**Please state your issue using the following template and, most importantly, in English.**

**Describe the bug**
```
[2020-10-21 20:41:52-ERROR-milvus.client.grpc_handler]: Error: <BaseException: (code=1, message=Invalid partition tag: (.*)tag. The first character of a partition tag must be an underscore or letter.)> (grpc_handler.py:44)
______ TestSearchBase.test_search_index_partitions_B[get_simple_index2-1] ______

self = <test_search.TestSearchBase object at 0x7f2f651f35f8>
connect = <milvus.client.stub.Milvus object at 0x7f2f674337f0>
collection = 'test_KHGVyHLT'
get_simple_index = {'index_type': 'IVF_SQ8', 'metric_type': 'L2', 'params': {'nlist': 128}}
get_top_k = 1

    @pytest.mark.level(2)
    def test_search_index_partitions_B(self, connect, collection, get_simple_index, get_top_k):
        '''
        target: test basic search function, all the search params is corrent, test all index params, and build
        method: search collection with the given vectors and tags, check the result
        expected: the length of the result is top_k
        '''
        top_k = get_top_k
        nq = 2
        tag = "tag"
        new_tag = "new_tag"
        index_type = get_simple_index["index_type"]
        if index_type in skip_pq():
            pytest.skip("Skip PQ")
        connect.create_partition(collection, tag)
        connect.create_partition(collection, new_tag)
        entities, ids = init_data(connect, collection, partition_tags=tag)
        new_entities, new_ids = init_data(connect, collection, nb=6001, partition_tags=new_tag)
        connect.create_index(collection, field_name, get_simple_index)
        search_param = get_search_param(index_type)
        query, vecs = gen_query_vectors(field_name, new_entities, top_k, nq, search_params=search_param)
        if top_k > max_top_k:
            with pytest.raises(Exception) as e:
                res = connect.search(collection, query)
        else:
>           res = connect.search(collection, query, partition_tags=["(.*)tag"])

entity/test_search.py:388: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/local/lib/python3.6/site-packages/milvus/client/stub.py:33: in inner
    return func(self, *args, **kwargs)
/usr/local/lib/python3.6/site-packages/milvus/client/stub.py:557: in search
    return handler.search(collection_name, dsl, partition_tags, fields, timeout=timeout, **kwargs)
/usr/local/lib/python3.6/site-packages/milvus/client/grpc_handler.py:50: in handler
    raise e
/usr/local/lib/python3.6/site-packages/milvus/client/grpc_handler.py:42: in handler
    return func(self, *args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <milvus.client.grpc_handler.GrpcHandler object at 0x7f2f655656d8>
collection_name = 'test_KHGVyHLT'
query_entities = {'bool': {'must': [{'vector': {'float_vector': {'metric_type': 'L2', 'params': {...}, 'query': [...], 'topk': 1}}}]}}
partition_tags = ['(.*)tag'], fields = None, kwargs = {'timeout': None}
request = collection_name: "test_KHGVyHLT"
partition_tag_array: "(.*)tag"
vector_param {
  json: "{\"place_holder_0\":{\"float_v... "{\"bool\":{\"must\":[{\"vector\":\"place_holder_0\"}]}}"
extra_params {
  key: "params"
  value: "{\"fields\":[]}"
}

to = None
ft = <_MultiThreadedRendezvous of RPC that terminated with:
	status = StatusCode.OK
	details = ""
>
response = status {
  error_code: UNEXPECTED_ERROR
  reason: "Invalid partition tag: (.*)tag. The first character of a partition tag must be an underscore or letter."
}
entities {
}


    @error_handler(None)
    def search(self, collection_name, query_entities, partition_tags=None, fields=None, **kwargs):
        request = Prepare.search_param(collection_name, query_entities, partition_tags, fields)
        self._search_hook.pre_search()
        to = kwargs.get("timeout", None)
        ft = self._stub.Search.future(request, wait_for_ready=True, timeout=to)
        if kwargs.get("_async", False) is True:
            func = kwargs.get("_callback", None)
            return SearchFuture(ft, func)
    
        response = ft.result()
        self._search_hook.aft_search()
    
        if self._search_hook.on_response():
            return response
    
        if response.status.error_code != 0:
>           raise BaseException(response.status.error_code, response.status.reason)
E           milvus.client.exceptions.BaseException: <BaseException: (code=1, message=Invalid partition tag: (.*)tag. The first character of a partition tag must be an underscore or letter.)>

/usr/local/lib/python3.6/site-packages/milvus/client/grpc_handler.py:479: BaseException
----------------------------- Captured stderr call -----------------------------
Error: <BaseException: (code=1, message=Invalid partition tag: (.*)tag. The first character of a partition tag must be an underscore or letter.)>
------------------------------ Captured log call -------------------------------
```
**Steps/Code to reproduce behavior**
http://internal.zilliz.com:18080/jenkins/job/milvus-ci/job/master/1275/artifact/_artifacts/centos7/cpu/milvus_cpu_mysql_dev_test.log
**Expected behavior**
A clear and concise description of what you expected to happen.

**Method of installation**
- [ ] Docker/cpu
- [ ] Docker/gpu
- [ ] Build from source

**Environment details**
- Hardware/Softwars conditions (OS, CPU, GPU, Memory)


- Milvus version (master or released version)

master bracch


**Configuration file**
Settings you made in `server_config.yaml` or `milvus.yaml`

```yaml
paste-file-content-here
```

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Additional context**
Add any other context about the problem here.

--- a/core/src/db/DBImpl.cpp
+++ b/core/src/db/DBImpl.cpp
@@ -19,8 +19,6 @@
 #include "db/merge/MergeManagerFactory.h"
 #include "db/merge/MergeTask.h"
 #include "db/snapshot/CompoundOperations.h"
-#include "db/snapshot/EventExecutor.h"
-#include "db/snapshot/OperationExecutor.h"
 #include "db/snapshot/ResourceHelper.h"
 #include "db/snapshot/ResourceTypes.h"
 #include "db/snapshot/Snapshots.h"
@@ -32,16 +30,13 @@
 #include "scheduler/SchedInst.h"
 #include "scheduler/job/SearchJob.h"
 #include "segment/SegmentReader.h"
-#include "segment/SegmentWriter.h"
 #include "segment/Utils.h"
-#include "server/ValidationUtil.h"
 #include "utils/Exception.h"
 #include "utils/TimeRecorder.h"
 
 #include <fiu/fiu-local.h>
 #include <src/scheduler/job/BuildIndexJob.h>
 #include <algorithm>
-#include <functional>
 #include <limits>
 #include <unordered_set>
 #include <utility>
@@ -319,7 +314,6 @@ DBImpl::HasPartition(const std::string& collection_name, const std::string& part
     CHECK_AVAILABLE
 
     snapshot::ScopedSnapshotT ss;
-    STATUS_CHECK(server::ValidatePartitionTags({partition_tag}));
     STATUS_CHECK(snapshot::Snapshots::GetInstance().GetSnapshot(ss, collection_name));
 
     auto partition_tags = std::move(ss->GetPartitionNames());

--- a/core/src/scheduler/selector/FaissIVFSQ8HPass.cpp
+++ b/core/src/scheduler/selector/FaissIVFSQ8HPass.cpp
@@ -21,6 +21,8 @@
 #include "server/ValidationUtil.h"
 #include "utils/Log.h"
 
+#include <utility>
+
 namespace milvus {
 namespace scheduler {
 

--- a/core/src/scheduler/selector/FaissIVFSQ8HPass.cpp
+++ b/core/src/scheduler/selector/FaissIVFSQ8HPass.cpp
@@ -24,6 +24,19 @@
 namespace milvus {
 namespace scheduler {
 
+namespace {
+bool
+SpecifyToCPU(milvus::json json) {
+    try {
+        // if the 'nprobe' field is missed or empty value, an exception will be thrown
+        auto nprobe = json[knowhere::IndexParams::nprobe].get<int64_t>();
+        return nprobe > server::GPU_QUERY_MAX_NPROBE;
+    } catch (std::exception& e) {
+        return true;
+    }
+}
+}  // namespace
+
 FaissIVFSQ8HPass::FaissIVFSQ8HPass() {
     ConfigMgr::GetInstance().Attach("gpu.gpu_search_threshold", this);
 }
@@ -58,7 +71,7 @@ FaissIVFSQ8HPass::Run(const TaskPtr& task) {
     } else if (search_task->topk() > milvus::server::GPU_QUERY_MAX_TOPK) {
         LOG_SERVER_DEBUG_ << LogOut("FaissIVFSQ8HPass: topk > gpu_max_topk_threshold, specify cpu to search!");
         res_ptr = ResMgrInst::GetInstance()->GetResource("cpu");
-    } else if (search_task->ExtraParam()[knowhere::IndexParams::nprobe].get<int64_t>() > server::GPU_QUERY_MAX_NPROBE) {
+    } else if (SpecifyToCPU(std::move(search_task->ExtraParam()))) {
         LOG_SERVER_DEBUG_ << LogOut("FaissIVFSQ8HPass: nprobe > gpu_max_nprobe_threshold, specify cpu to search!");
         res_ptr = ResMgrInst::GetInstance()->GetResource("cpu");
     } else if (search_task->nq() < (uint64_t)threshold_) {

--- a/core/src/server/ValidationUtil.cpp
+++ b/core/src/server/ValidationUtil.cpp
@@ -90,16 +90,21 @@ CheckParameterExistence(const milvus::json& json_params, const std::string& para
 
 Status
 ValidateCollectionName(const std::string& collection_name) {
+    // trim side-blank of tag, only compare valid characters
+    // for example: " ab cd " is treated as "ab cd"
+    std::string valid_name = collection_name;
+    StringHelpFunctions::TrimStringBlank(valid_name);
+
     // Collection name shouldn't be empty.
-    if (collection_name.empty()) {
+    if (valid_name.empty()) {
         std::string msg = "Collection name should not be empty.";
         LOG_SERVER_ERROR_ << msg;
         return Status(SERVER_INVALID_COLLECTION_NAME, msg);
     }
 
-    std::string invalid_msg = "Invalid collection name: " + collection_name + ". ";
+    std::string invalid_msg = "Invalid collection name: " + valid_name + ". ";
     // Collection name size shouldn't exceed engine::MAX_NAME_LENGTH.
-    if (collection_name.size() > engine::MAX_NAME_LENGTH) {
+    if (valid_name.size() > engine::MAX_NAME_LENGTH) {
         std::string msg = invalid_msg + "The length of a collection name must be less than " +
                           std::to_string(engine::MAX_NAME_LENGTH) + " characters.";
         LOG_SERVER_ERROR_ << msg;
@@ -107,16 +112,16 @@ ValidateCollectionName(const std::string& collection_name) {
     }
 
     // Collection name first character should be underscore or character.
-    char first_char = collection_name[0];
+    char first_char = valid_name[0];
     if (first_char != '_' && std::isalpha(first_char) == 0) {
         std::string msg = invalid_msg + "The first character of a collection name must be an underscore or letter.";
         LOG_SERVER_ERROR_ << msg;
         return Status(SERVER_INVALID_COLLECTION_NAME, msg);
     }
 
-    int64_t table_name_size = collection_name.size();
+    int64_t table_name_size = valid_name.size();
     for (int64_t i = 1; i < table_name_size; ++i) {
-        char name_char = collection_name[i];
+        char name_char = valid_name[i];
         if (name_char != '_' && name_char != '$' && std::isalnum(name_char) == 0) {
             std::string msg = invalid_msg + "Collection name can only contain numbers, letters, and underscores.";
             LOG_SERVER_ERROR_ << msg;
@@ -454,60 +459,50 @@ ValidateSearchTopk(int64_t top_k) {
 }
 
 Status
-ValidatePartitionTags(const std::vector<std::string>& partition_tags) {
+ValidatePartitionTags(const std::vector<std::string>& partition_tags, bool for_create) {
     for (const std::string& tag : partition_tags) {
-        // Partition nametag shouldn't be empty.
-        if (tag.empty()) {
-            std::string msg = "Partition tag should not be empty.";
-            LOG_SERVER_ERROR_ << msg;
-            return Status(SERVER_INVALID_PARTITION_TAG, msg);
-        }
-
-        std::string invalid_msg = "Invalid partition tag: " + tag + ". ";
-        // Partition tag size shouldn't exceed 255.
-        if (tag.size() > engine::MAX_NAME_LENGTH) {
-            std::string msg = invalid_msg + "The length of a partition tag must be less than 255 characters.";
-            LOG_SERVER_ERROR_ << msg;
-            return Status(SERVER_INVALID_PARTITION_TAG, msg);
-        }
-
-        // Partition tag first character should be underscore or character.
-        char first_char = tag[0];
-        if (first_char != '_' && std::isalnum(first_char) == 0) {
-            std::string msg = invalid_msg + "The first character of a partition tag must be an underscore or letter.";
-            LOG_SERVER_ERROR_ << msg;
-            return Status(SERVER_INVALID_PARTITION_TAG, msg);
-        }
-
-        int64_t tag_size = tag.size();
-        for (int64_t i = 1; i < tag_size; ++i) {
-            char name_char = tag[i];
-            if (name_char != '_' && name_char != '$' && std::isalnum(name_char) == 0) {
-                std::string msg = invalid_msg + "Partition tag can only contain numbers, letters, and underscores.";
-                LOG_SERVER_ERROR_ << msg;
-                return Status(SERVER_INVALID_PARTITION_TAG, msg);
-            }
-        }
-
-#if 0
         // trim side-blank of tag, only compare valid characters
         // for example: " ab cd " is treated as "ab cd"
         std::string valid_tag = tag;
         StringHelpFunctions::TrimStringBlank(valid_tag);
+
+        std::string invalid_msg = "Invalid partition tag: " + tag + ". ";
         if (valid_tag.empty()) {
-            std::string msg = "Invalid partition tag: " + valid_tag + ". " + "Partition tag should not be empty.";
+            std::string msg = invalid_msg + "Partition tag should not be empty.";
             LOG_SERVER_ERROR_ << msg;
             return Status(SERVER_INVALID_PARTITION_TAG, msg);
         }
 
         // max length of partition tag
         if (valid_tag.length() > engine::MAX_NAME_LENGTH) {
-            std::string msg = "Invalid partition tag: " + valid_tag + ". " +
-                              "Partition tag exceed max length: " + std::to_string(engine::MAX_NAME_LENGTH);
+            std::string msg = invalid_msg + "The length of a partition tag must be less than " +
+                              std::to_string(engine::MAX_NAME_LENGTH) + " characters.";
             LOG_SERVER_ERROR_ << msg;
             return Status(SERVER_INVALID_PARTITION_TAG, msg);
         }
-#endif
+
+        // for CreatePartition, a partition tag must be numbers, letters, and underscores
+        // but for Search/HasPartition, user can check partition by regex pattern, such as p_(.*) match p_1, p_50...
+        if (for_create) {
+            // Partition tag first character should be underscore or character.
+            char first_char = tag[0];
+            if (first_char != '_' && std::isalnum(first_char) == 0) {
+                std::string msg =
+                    invalid_msg + "The first character of a partition tag must be an underscore or letter.";
+                LOG_SERVER_ERROR_ << msg;
+                return Status(SERVER_INVALID_PARTITION_TAG, msg);
+            }
+
+            int64_t tag_size = tag.size();
+            for (int64_t i = 1; i < tag_size; ++i) {
+                char name_char = tag[i];
+                if (name_char != '_' && name_char != '$' && std::isalnum(name_char) == 0) {
+                    std::string msg = invalid_msg + "Partition tag can only contain numbers, letters, and underscores.";
+                    LOG_SERVER_ERROR_ << msg;
+                    return Status(SERVER_INVALID_PARTITION_TAG, msg);
+                }
+            }
+        }
     }
 
     return Status::OK();

--- a/core/src/server/ValidationUtil.h
+++ b/core/src/server/ValidationUtil.h
@@ -57,7 +57,7 @@ extern Status
 ValidateSearchTopk(int64_t top_k);
 
 extern Status
-ValidatePartitionTags(const std::vector<std::string>& partition_tags);
+ValidatePartitionTags(const std::vector<std::string>& partition_tags, bool for_create);
 
 extern Status
 ValidateInsertDataSize(const InsertParam& insert_param);

--- a/core/src/server/delivery/request/CreatePartitionReq.cpp
+++ b/core/src/server/delivery/request/CreatePartitionReq.cpp
@@ -48,7 +48,7 @@ CreatePartitionReq::OnExecute() {
             return Status(SERVER_INVALID_PARTITION_TAG, "'_default' is built-in partition tag");
         }
 
-        auto status = ValidatePartitionTags({tag_});
+        auto status = ValidatePartitionTags({tag_}, true);
         fiu_do_on("CreatePartitionReq.OnExecute.invalid_partition_tags",
                   status = Status(milvus::SERVER_UNEXPECTED_ERROR, ""));
         if (!status.ok()) {

--- a/core/src/server/delivery/request/HasPartitionReq.cpp
+++ b/core/src/server/delivery/request/HasPartitionReq.cpp
@@ -46,6 +46,7 @@ HasPartitionReq::OnExecute() {
 
         // step 1: check arguments
         STATUS_CHECK(ValidateCollectionName(collection_name_));
+        STATUS_CHECK(ValidatePartitionTags({partition_tag_}, true));
 
         bool exists = false;
         STATUS_CHECK(DBWrapper::DB()->HasCollection(collection_name_, exists));

--- a/core/src/server/delivery/request/SearchReq.cpp
+++ b/core/src/server/delivery/request/SearchReq.cpp
@@ -55,7 +55,7 @@ SearchReq::OnExecute() {
         TimeRecorderAuto rc(hdr);
 
         STATUS_CHECK(ValidateCollectionName(query_ptr_->collection_id));
-        STATUS_CHECK(ValidatePartitionTags(query_ptr_->partitions));
+        STATUS_CHECK(ValidatePartitionTags(query_ptr_->partitions, false));
 
         // step 2: check table existence
         // only process root table, ignore partition table

--- a/core/src/utils/StringHelpFunctions.cpp
+++ b/core/src/utils/StringHelpFunctions.cpp
@@ -10,6 +10,7 @@
 // or implied. See the License for the specific language governing permissions and limitations under the License.
 
 #include "utils/StringHelpFunctions.h"
+#include "utils/Log.h"
 
 #include <fiu/fiu-local.h>
 #include <algorithm>
@@ -146,9 +147,16 @@ StringHelpFunctions::IsRegexMatch(const std::string& target_str, const std::stri
     }
 
     // regex match
-    std::regex pattern(pattern_str);
-    std::smatch results;
-    return std::regex_match(target_str, results, pattern);
+    // for illegal regex expression, the std::regex will throw exception, regard as unmatch
+    try {
+        std::regex pattern(pattern_str);
+        std::smatch results;
+        return std::regex_match(target_str, results, pattern);
+    } catch (std::exception& e) {
+        LOG_SERVER_ERROR_ << "Regex exception: " << e.what();
+    }
+
+    return false;
 }
 
 Status

